flight/Modules/Actuator/actuator.c

```diff
@@ -1058,6 +1108,73 @@ static void set_failsafe()
 	ActuatorCommandChannelSet(Channel);
 }
 
+static struct smith_predictor* smithp_init(float delay, float iir, float mix)
+{
+	struct smith_predictor *m = PIOS_malloc_no_dma(sizeof(*m));
+	if (!m)
+		return NULL;
+	memset(m, 0, sizeof(*m));
+
+	if (delay > SMITHP_MAX_DELAY) delay = SMITHP_MAX_DELAY;
+
+	m->samples = delay;
+	m->iir_alpha = iir;
+	m->mix = mix;
+
+	m->data = PIOS_malloc_no_dma(delay * MAX_MIX_ACTUATORS * sizeof(*m->data));
+	if (!m->data)
+		return NULL;
+	memset(m->data, 0, sizeof(*m->data) * delay * MAX_MIX_ACTUATORS);
+
+	return m;
+}
+
+/*
+	Call order for it to work should be:
+	- Compensate
+	- Push vect
+*/
+
+static void smithp_push_vect(struct smith_predictor *m, float *motor_vect)
+{
+	if (!m) return;
+
+	uint32_t p = (m->idx % m->samples) * MAX_MIX_ACTUATORS;
+
+	/* Write motor vector to position. */
+	for (int i = 0; i < MAX_MIX_ACTUATORS; i++) {
+		float v = motor_vect[i];
+		m->data[p++] = v != v ? 0 : v;
+	}
+
+	/* Moves pointer to oldest data point (via modulo) for _compensate. */
+	m->idx++;
+}
+
+static void smithp_compensate(struct smith_predictor *m, float *desired_vect)
+{
+	if (!m || !inverse_mixer_computed) return;
+
+	uint32_t p = (m->idx % m->samples) * MAX_MIX_ACTUATORS;
+
+	float inv[MIXERSETTINGS_MIXER1VECTOR_NUMELEM];
+
+	for (int i = 0; i < MAX_MIX_ACTUATORS; i++) {
+		/* Do IIR on delayed data*/
+		m->iir[i] = m->iir[i] * m->iir_alpha + (1 - m->iir_alpha) * m->data[p++];
+	}
+
+	/* Do inverse stuff. */
+	matrix_mul_check(motor_mixer_inv, m->iir, inv,
+			MIXERSETTINGS_MIXER1VECTOR_NUMELEM,
+			MAX_MIX_ACTUATORS,
+			1);
+
+	for (int i = 0; i < MIXERSETTINGS_MIXER1VECTOR_NUMELEM; i++) {
+		desired_vect[i] += m->mix * (inv[i] - desired_vect[i]);
```

Comment:
I'm hoping this one is the right way. I had the terms of the substraction switched earlier and it made things audibly worse.